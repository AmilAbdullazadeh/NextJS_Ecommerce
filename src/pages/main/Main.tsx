import React, { useEffect } from 'react';
import Alert from '@material-ui/lab/Alert';

import { ApiPhone } from '@api';
import { useAppDispatch } from '@hooks';
import { setRunningAction } from '@store/actions';
import { Typography, ProductCard } from '@components';

import { Head, Jumbotron } from './components';
import useStyles from './Main.styles';

export type Phone = Pick<
    ApiPhone,
    | 'id'
    | 'name'
    | 'ratingValue'
    | 'reviewCount'
    | 'price'
    | 'availability'
    | 'description'
    | 'image'
>;

export interface PhoneData {
    results: Phone[];
}

export interface MainProps {
    frontPagePhones?: PhoneData;
}

const Main: React.FC<MainProps> = ({ frontPagePhones }: MainProps) => {
    const classes = useStyles();
    const dispatch = useAppDispatch();

    useEffect(() => {
        dispatch(setRunningAction({ running: true }));
    }, []);

    return (
        <>
            <Head
                productCollectionSize={frontPagePhones?.results.length ?? 0}
            />
            <Jumbotron />
            {!frontPagePhones && (
                <Alert severity="error">
                    Error fetching top selling phones
                </Alert>
            )}
            <section className={classes.phonesContainer}>
                {frontPagePhones && (
                    <>
                        <Typography
                            variant="h6"
                            component="h3"
                            className={classes.topSellingPhonesText}
                        >
                            Top selling smartphones
                        </Typography>
                        {frontPagePhones.results.map(phone => (
                            <ProductCard
                                key={phone.id}
                                id={phone.id}
                                name={phone.name}
                                ratingValue={phone.ratingValue ?? 0}
                                reviewCount={phone.reviewCount ?? 0}
                                price={phone.price}
                                availability={phone.availability}
                                description={phone.description}
                                image={phone.image}
                                imageFallback={'phoneFallBack.png'}
                            />
                        ))}
                    </>
                )}
            </section>
        </>
    );
};

export default Main;
