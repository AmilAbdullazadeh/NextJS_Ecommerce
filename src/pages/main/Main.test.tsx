import React from 'react';

import { renderWithProviders } from '@testUtils';

import Main, { MainProps } from './Main';

describe('Main page', () => {
    const defaultProps: MainProps = {
        frontPagePhones: {
            results: [
                {
                    id: 'mock id',
                    name: 'mock name',
                    ratingValue: 0,
                    reviewCount: 0,
                    price: 240,
                    availability: 4,
                    description: 'mock description',
                    image: 'image src',
                },
            ],
        },
    };

    it('renders correctly', () => {
        const { container } = renderWithProviders(<Main {...defaultProps} />);
        expect(container.firstChild).toMatchSnapshot();
    });

    it('renders correctly for no phones', () => {
        const noPhonesProps: MainProps = { frontPagePhones: { results: [] } };
        const { container } = renderWithProviders(<Main {...noPhonesProps} />);
        expect(container.firstChild).toMatchSnapshot();
    });

    it('renders alert when the phones prop is undefined', () => {
        const { container } = renderWithProviders(<Main />);
        expect(container.firstChild).toMatchSnapshot();
    });

    it('renders correctly for no rating', () => {
        const noRatingProps: MainProps = {
            frontPagePhones: {
                results: [
                    {
                        id: 'mock id',
                        name: 'mock name',
                        ratingValue: undefined,
                        reviewCount: undefined,
                        price: 240,
                        availability: 4,
                        description: 'mock description',
                        image: 'image src',
                    },
                ],
            },
        };
        const { container } = renderWithProviders(<Main {...noRatingProps} />);
        expect(container.firstChild).toMatchSnapshot();
    });
});
