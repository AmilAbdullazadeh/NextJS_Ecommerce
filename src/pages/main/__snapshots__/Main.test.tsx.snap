// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main page renders alert when the phones prop is undefined 1`] = `
<div
  class="MuiContainer-root makeStyles-container-38"
>
  <div
    class="MuiBox-root MuiBox-root-42"
  >
    <div
      class="MuiBox-root MuiBox-root-43"
    >
      <h2
        class="MuiTypography-root makeStyles-productName-39 MuiTypography-h2"
      >
        iPhone 13 Pro
      </h2>
    </div>
    <div
      class="MuiBox-root MuiBox-root-44"
    >
      <h2
        class="MuiTypography-root makeStyles-productMoto-40 MuiTypography-h4"
      >
        Oh.So.Pro
      </h2>
    </div>
  </div>
  <div
    class="MuiBox-root MuiBox-root-45 makeStyles-imageContainer-41"
  >
    <div
      style="display: block; overflow: hidden; position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; margin: 0px;"
    >
      <noscript />
      <img
        alt="iphone 13 Pro"
        data-nimg="true"
        decoding="async"
        src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
        style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%; object-fit: contain;"
      />
    </div>
  </div>
</div>
`;

exports[`Main page renders correctly 1`] = `
<div
  class="MuiContainer-root makeStyles-container-3"
>
  <div
    class="MuiBox-root MuiBox-root-7"
  >
    <div
      class="MuiBox-root MuiBox-root-8"
    >
      <h2
        class="MuiTypography-root makeStyles-productName-4 MuiTypography-h2"
      >
        iPhone 13 Pro
      </h2>
    </div>
    <div
      class="MuiBox-root MuiBox-root-9"
    >
      <h2
        class="MuiTypography-root makeStyles-productMoto-5 MuiTypography-h4"
      >
        Oh.So.Pro
      </h2>
    </div>
  </div>
  <div
    class="MuiBox-root MuiBox-root-10 makeStyles-imageContainer-6"
  >
    <div
      style="display: block; overflow: hidden; position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; margin: 0px;"
    >
      <noscript />
      <img
        alt="iphone 13 Pro"
        data-nimg="true"
        decoding="async"
        src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
        style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%; object-fit: contain;"
      />
    </div>
  </div>
</div>
`;

exports[`Main page renders correctly for no phones 1`] = `
<div
  class="MuiContainer-root makeStyles-container-28"
>
  <div
    class="MuiBox-root MuiBox-root-32"
  >
    <div
      class="MuiBox-root MuiBox-root-33"
    >
      <h2
        class="MuiTypography-root makeStyles-productName-29 MuiTypography-h2"
      >
        iPhone 13 Pro
      </h2>
    </div>
    <div
      class="MuiBox-root MuiBox-root-34"
    >
      <h2
        class="MuiTypography-root makeStyles-productMoto-30 MuiTypography-h4"
      >
        Oh.So.Pro
      </h2>
    </div>
  </div>
  <div
    class="MuiBox-root MuiBox-root-35 makeStyles-imageContainer-31"
  >
    <div
      style="display: block; overflow: hidden; position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; margin: 0px;"
    >
      <noscript />
      <img
        alt="iphone 13 Pro"
        data-nimg="true"
        decoding="async"
        src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
        style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%; object-fit: contain;"
      />
    </div>
  </div>
</div>
`;

exports[`Main page renders correctly for no rating 1`] = `
<div
  class="MuiContainer-root makeStyles-container-48"
>
  <div
    class="MuiBox-root MuiBox-root-52"
  >
    <div
      class="MuiBox-root MuiBox-root-53"
    >
      <h2
        class="MuiTypography-root makeStyles-productName-49 MuiTypography-h2"
      >
        iPhone 13 Pro
      </h2>
    </div>
    <div
      class="MuiBox-root MuiBox-root-54"
    >
      <h2
        class="MuiTypography-root makeStyles-productMoto-50 MuiTypography-h4"
      >
        Oh.So.Pro
      </h2>
    </div>
  </div>
  <div
    class="MuiBox-root MuiBox-root-55 makeStyles-imageContainer-51"
  >
    <div
      style="display: block; overflow: hidden; position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; margin: 0px;"
    >
      <noscript />
      <img
        alt="iphone 13 Pro"
        data-nimg="true"
        decoding="async"
        src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
        style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%; object-fit: contain;"
      />
    </div>
  </div>
</div>
`;
